onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -594,141 +588,149 @@ static Status GetFileContent(
   return Status::OK();
 }
 
-static void MoveOrtCallback(OrtCallback& from, OrtCallback& to) {
-  to.f = from.f;
-  to.param = from.param;
-  from.f = nullptr;
-  from.param = nullptr;
-}
-
 #define CASE_PROTO(X, Y)                                                      \
   case ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_##X:        \
     ORT_RETURN_IF_ERROR(                                                      \
         UnpackTensor<Y>(tensor_proto, raw_data, raw_data_len,                 \
                         (Y*)preallocated, static_cast<size_t>(tensor_size))); \
     break;
 
+/**
+ * @brief Convert tensor_proto to tensor format and store it to pre-allocated tensor 
+ * @param env 
+ * @param model_path     
+ * @param tensor_proto  tensor data in protobuf format
+ * @param tensorp       pre-allocated tensor object, where we store the data
+ * @return 
+*/
+Status TensorProtoToTensor(const Env& env, const ORTCHAR_T* model_path,
```

Comment:
The callback parameter in the original code is crucial. 
Without it, you need to define an AutoDelete object here and copy the tensor buffer, which means in some special cases it will double the peak memory usage.

Reply:
can you elaborate on which cases?

Reply:
this part went through several revisions and discussions with folks. I think I covered all cases. Currently only external file read may allocated extra memory sometimes but it's auto released by the destructor of deleter_for_file_data.  Please let me know if you suspect a specific scenario where memory is doubled.

Reply:
One of onnxruntime's major goal is keeping memory usage low. But we are not doing a very good job here. For example, if the model is 500MB large, we may need 1.5GB or more memory to load it. It is partially because protobuf is not designed for handling large messages.  It uses a vector to store the data read from the file.  Like
```
std::vector vec;
while(file has data){
    If the vector still has space
       read file data into vector
    else
       Double the vector's capacity // Here assume the current one is 500MB, then it will create a new vector with 1GB and copy the old one over it.
}
```
We can't **improve** protobuf,  so the best thing we can rely on is using "external files" to store the weights. And we want to avoid a tensor has two copies in the memory. And when Yuan designed the Tensor class, he thought it should be lightweight and shouldn't own memory. Memory buffers are either allocated by session state(or planner) or feed in from outside.  If we let the Tensor class own buffers, then we need to carefully track its lifetime so that we don't leak memory.  

```

Reply:
as described above, in the case of external file, this function may allocate a read buffer but it would be released by the destructor of deleter_for_file_data. So there is no doubling memory. 

Sorry I still don't see how this function may result in doubling of memory.

Reply:
actually this change FIXED the memory doubling issue in this function. Before this change, when double buffer happens, the extra buffer was passed out by the call back parameter you mentioned. Those parameters are held by the session and won't be released until session end. So in fact, those call back parameters are the source of memory doubling and this change fixed this issue.